features/323-login-id/type.md

```diff
@@ -0,0 +1,94 @@
+# Login ID Type
+
+## Background
+
+Skygear Auth supports login with login id and password. This document defines the types of login id.
+
+## Validation and Normalization
+
+Validation and normalization are defined base on login id type. When user signup with login ids, Skygear Auth will validate the login ids and reject the signup if it won't pass. After passing the validation, Skygear Auth will store the normalized login id for login.
+
+When user login with login id, login id key is optional. If login id key is provided, Skygear Auth will normalize the input by type, query the principal and perform the login. If login id key is not provided, Skygear Auth will walk through all login id keys, normalize the input and query the principal. If multiple principals are found, the login process will be stopped and error will be returned.
+
+## Types
+
+This section defines the login id types and its validation and normalization rules.
+
+- [Email](#Email)
+- [Username](#Username)
+- [Phone](#Phone)
+- [Raw](#Raw)
+
+### Email
+
+#### Validation
+
+- [RFC 5322 address](https://tools.ietf.org/html/rfc5322#section-3.4.1) (Compulsory)
+
+#### Normalization
+
+- Change domain part to lower case (Compulsory)
+- Change local part to lower case (Configurable, default ON)
+- Remove words after `+` sign in the local part (Configurable, default OFF)
+- Remove all `.` sign in the local part (Configurable, default OFF)
+
+#### Configuration
+```yaml
+user_config:
+  auth:
+    login_id:
```

Comment:
I think using `login_id` as the config key is confusing; maybe call it `login_id_handling`?

Reply:
login_id_handling is too verbose IMO

Reply:
Cannot think of any better name which is shorter, I changed it to `login_id_handling`. See if we have any other ideas? Otherwise we will use this one.

Reply:
What about `login_id_format`?

Reply:
I think it is more than just formatting? e.g. we reject username from list of keywords

Reply:
My only concern is I haven't seen people using the term `handling` in the configuration name.

Reply:
Maybe use `login_id_types`?